### PR TITLE
Run building simulations on GCP Batch

### PR DESCRIPTION
**Pull request by nweires**
**This change:**
- Before starting the batch job (`run_batch()`):
  - Run sampler to generate buildings to simulate
  - Gather, compress, and upload various required files to GCS bucket (with some extra logic to avoid uploading duplicate weather files)
  - Split samples into batches of buildings, with a JSON file describing each batch
- During each task (`run_task()`):
  -  Collect and extract/decompress files needed for this particular task / set of buildings (with extra logic to fetch only the weather files needed)
  - Run OpenStudio to simulate each building
  - Write output files back to the GCS bucket

Note: Most of this code is copied from aws.py. The main differences are around how file uploads are handled on S3 vs GCS. In particular, GCP batch makes it easy to mount a bucket so it can be treated like a normal directory, instead of explicitly downloading files from the bucket.

**Next steps:**
- Clean up configuration options and defaults

**Tested:**
- Did a test run with 1000 samples in 20 batches, running on `ec-highcpu-4` machines. ([here](https://console.cloud.google.com/batch/jobsDetail/regions/us-central1/jobs/natalie10023-10-27-022258/details?project=buildstockbatch-dev)) It ran for just under an hour on 5 machines and succeeded. The output files are [here](https://console.cloud.google.com/storage/browser/testing-natalie/test_1000_samples?project=buildstockbatch-dev).

### CODE REVIEW

File: buildstockbatch/gcp/gcp.py [hunk 1]
```diff
@@ -56,6 +72,50 @@ def docker_image(self):
...
+    transfer_manager.upload_many_from_filenames(
+        bucket,
+        string_paths,
+        source_directory=local_dir_abs,
+        blob_name_prefix=prefix,
+        raise_exception=True,
+    )
+
+
+# todo: aws-shared (see file comment)
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py:104]: To reduce code duplication (and a half-step towards refactoring), can we just call `buildstockbatch.aws.compress_file()` from here? If so, ditto for `calc_hash_for_file()`.

🤔 And then I guess — not for this PR, necessarily — similarly we could also have GcpBatch inherit from AWS's DockerBatchBase (i.e., `class GcpBatch(buildstockbatch.aws.DockerBatchBase)`)?

File: buildstockbatch/gcp/gcp.py [hunk 2]
```diff
@@ -56,6 +72,50 @@ def docker_image(self):
...
+def calc_hash_for_file(filename):
+    with open(filename, 'rb') as f:
+        return hashlib.sha256(f.read()).hexdigest()
+
+
+def copy_GCS_file(src_bucket, src_name, dest_bucket, dest_name):
+    storage_client = storage.Client()
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py:118]: Any idea if there's any significant overhead in creating a new Client every time and whether we should consider passing a shared one in, instead?

File: buildstockbatch/gcp/gcp.py [hunk 3]
```diff
@@ -79,12 +139,18 @@ def __init__(self, project_filename):
...
+        self.batch_array_size = self.cfg['gcp']['batch_array_size']
 
         # Add timestamp to job ID, since duplicates aren't allowed (even between finished and new jobs)
         # TODO: stop appending timestamp here - it's useful for testing, but we should probably
         # make users choose a unique job ID each time.
         self.unique_job_id = self.job_identifier + datetime.utcnow().strftime('%y-%m-%d-%H%M%S')
 
+    # todo: aws-shared (see file comment)
+    @property
+    def weather_dir(self):
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: Can we put this below `def docker_image` like it is in aws.py?

File: buildstockbatch/gcp/gcp.py [hunk 4]
```diff
@@ -270,50 +336,164 @@ def run_batch(self):
...
+                    json.dump(
+                        {
+                            'job_num': i,
+                            'n_datapoints': n_datapoints,
+                            'batch': batch,
+                        },
+                        f,
+                        indent=4,
+                    )
+            task_count = i
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py:463]: While an improvement, this makes it slightly more complex to refactor to share with AWS, so I'd discourage diverging.

File: buildstockbatch/gcp/gcp.py [hunk 5]
```diff
@@ -270,50 +336,164 @@ def run_batch(self):
             - kick off a batch simulation on GCP
         """
         # Step 1: Run sampling and split up buildings into batches.
-        # TDOO: Generate buildstock.csv
-        # buildstock_csv_filename = self.sampler.run_sampling()
+        buildstock_csv_filename = self.sampler.run_sampling()
+
+        # Step 2: Compress and upload weather data and any other required files to GCP
+        # todo: aws-shared (see file comment)
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py:376]: FYI, I didn't look at most of this code block closely since it's mostly identical to the existing aws code.

File: buildstockbatch/gcp/gcp.py [hunk 6]
```diff
@@ -270,50 +336,164 @@ def run_batch(self):
...
+        bucket = self.gcs_bucket
+        Parallel(n_jobs=-1, verbose=9)(delayed(copy_GCS_file)(bucket, src, bucket, dest) for src, dest in epws_to_copy)
+
+        # Create the output directories
+        storage_client = storage.Client()
+        bucket = storage_client.bucket(self.gcs_bucket)
+        for upgrade_id in range(len(self.cfg.get('upgrades', [])) + 1):
+            dir_name = f'{self.gcs_prefix}/results/simulation_output/timeseries/up{upgrade_id:02d}/'
+            blob = bucket.blob(dir_name)
+            # Note: GCS doesn't have real directories, so these are just empty files with names ending in '/'.
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: Oh, I think this may explain the issue I'm having with postprocessing where "listing the contents of a directory" ends up "including the 'directory' in the listing". So, I think we should avoid doing this, if possible. Do we need to "create the output directories" — can you just delete this block of code? That would probably allow me to eliminate some changes I had to make to postprocessing.py (including some "If GCP, list files this way instead" code)

File: buildstockbatch/gcp/gcp.py [hunk 7]
```diff
@@ -338,13 +516,15 @@ def run_batch(self):
         # TODO: look into best default machine type for running OpenStudio, but also allow
         # changing via the project config. https://cloud.google.com/compute/docs/machine-types
         policy = batch_v1.AllocationPolicy.InstancePolicy(
+            # TODO: Skip setting this and let GCP pick the right machine type
+            # based on resourcesin ComputeResource
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: "resourcesin" → "resources in"

File: buildstockbatch/gcp/gcp.py [hunk 8]
```diff
@@ -338,13 +516,15 @@ def run_batch(self):
...
             machine_type='e2-standard-2',
+            provisioning_model=batch_v1.AllocationPolicy.ProvisioningModel.SPOT
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: This syntax+indentation looks weird and hard to parse, imho / easy to misinterpret. Maybe wrap it in parens to make it clearer?

e.g.,
```
 provisioning_model= (
    batch_v1.AllocationPolicy.ProvisioningModel.SPOT if self.use_spot
    else batch_v1.AllocationPolicy.ProvisioningModel.STANDARD),
```

File: buildstockbatch/gcp/gcp.py [hunk 9]
```diff
@@ -379,21 +559,165 @@ def run_batch(self):
...
         results back to GCS.
+
+        :param task_index: Index of this task (e.g. this may be task 1 of 4)
+        :param job_name: Job identifier
+        :param gcs_prefix: Prefix used for GCS files
         """
-        pass
+        # Local directory where we'll write files
+        sim_dir = pathlib.Path('/var/simdata/openstudio')
+        # Mounted GCS bucket directory
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: How/why did you choose to mount rather than download/upload via storage client?

One reason for the question is for sharing with AWS — how should we abstract/extract file access (this does seem like a good way).

Another is performance — is copy via mount any different than copying via API?

And then there's creating the fake directory (another comment) — did this approach make that necessary? If so, perhaps "create the directory" here (via filesystem API rather than the hacky way)?

File: buildstockbatch/gcp/gcp.py [hunk 10]
```diff
@@ -379,21 +559,165 @@ def run_batch(self):
...
+        # Look through the buildstock.csv to find the appropriate location and epw
+        epws_to_download = set()
+        building_ids = [x[0] for x in jobs_d['batch']]
+        with open(sim_dir / 'lib' / 'housing_characteristics' / 'buildstock.csv', 'r', encoding='utf-8') as f:
+            csv_reader = csv.DictReader(f)
+            for row in csv_reader:
+                if int(row['Building']) in building_ids:
+                    epws_to_download.add(epws_by_option[row[param_name]])
+
+        # Copy files to local machine and unzip the epws needed for these simulations
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: Isn't this just extracting from the mounted dir to the local machine (i.e., not actually _copying_ the file to local before extracting)?

File: buildstockbatch/gcp/gcp.py [hunk 11]
```diff
@@ -379,21 +559,165 @@ def run_batch(self):
...
-        logger.debug(created_job)
         logger.info('Newly created GCP Batch job')
-        logger.info(f'Job name: {created_job.name}')
-        logger.info(f'Job UID: {created_job.uid}')
+        logger.info(f'  Job name: {created_job.name}')
+        logger.info(f'  Job UID: {created_job.uid}')
 
     @classmethod
-    def run_job(self, job_id, gcs_bucket, gcs_prefix, job_name, region):
+    def run_task(cls, task_index, job_name, gcs_prefix):
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: FYI, I didn't look at most of this code block closely since it's mostly identical to the existing aws code.

File: buildstockbatch/gcp/gcp.py [hunk 12]
```diff
@@ -425,13 +749,10 @@ def main():
...
         # If this var exists, we're inside a single batch task.
-        job_id = int(os.environ['BATCH_TASK_INDEX'])
-        # TODO: pass in these env vars to tasks as needed. (Do we really need bucket here?)
-        gcs_bucket = os.environ['GCS_BUCKET']
+        task_index = int(os.environ['BATCH_TASK_INDEX'])
         gcs_prefix = os.environ['GCS_PREFIX']
-        job_name = os.environ['JOB_NAME']
-        region = os.environ['REGION']
-        GcpBatch.run_job(job_id, gcs_bucket, gcs_prefix, job_name, region)
+        job_name = os.environ['JOB_ID']
```
Comment (lathanh) [inline buildstockbatch/gcp/gcp.py]: can the env-var stay 'JOB_NAME' to match the var it's fed into?

File: buildstockbatch/gcp/gcp.py (hunk 1 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py:104]: I was following your lead from PR #1 where you copied things in instead of importing them from aws.py. 🙂 There's enough here to refactor that I'd rather just save it all for one big PR at this point.

File: buildstockbatch/gcp/gcp.py (hunk 2 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py:118]: No, the constructor doesn't seem to do much at all. As a quick test, I passed in bogus credentials to see if it used them, and it doesn't complain until you try to actually read or write something.

File: buildstockbatch/gcp/gcp.py (hunk 3 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: Yep, done.

File: buildstockbatch/gcp/gcp.py (hunk 6 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: Removed.

File: buildstockbatch/gcp/gcp.py (hunk 8 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: Done

File: buildstockbatch/gcp/gcp.py (hunk 9 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: Switched to gcsfs, as discussed offline.

File: buildstockbatch/gcp/gcp.py (hunk 4 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py:463]: I found the existing variables names here confusing (even when comparing to the AWS docs), so I made a couple small changes that I'd like to use in the refactored version as well.

File: buildstockbatch/gcp/gcp.py (hunk 12 above)
Comment (nweires) [reply buildstockbatch/gcp/gcp.py]: Yep, done.

**Review by nweires — COMMENTED** [review]
Updated to use gcsfs instead of mounting the bucket.

File: buildstockbatch/gcp/gcp.py (hunk 1 above)
Comment (lathanh) [reply buildstockbatch/gcp/gcp.py:104]: Yea, I've learned a little bit since PR #1. refactoring later SGTM

**Review by lathanh — APPROVED** [review]
